engineio/transport/polling/encoder.go

```diff
@@ -0,0 +1,132 @@
+package polling
+
+import (
+	"encoding/base64"
+	"io"
+	"sync/atomic"
+	"time"
+
+	"github.com/googollee/go-socket.io/engineio/frame"
+)
+
+// encoder encodes frames to the writer.
+type encoder struct {
+	pingTimeout       time.Duration
+	lastPing          time.Time
+	writer            *bufWriter
+	hasFramesChan     chan struct{}
+	closed            chan struct{}
+	hasNonClosedFrame int32
+}
+
+func newEncoder(pingTimouet time.Duration, closed chan struct{}, buf []byte) *encoder {
+	return &encoder{
+		pingTimeout:   pingTimouet,
+		lastPing:      time.Now(),
+		writer:        newBufWriter(buf),
+		hasFramesChan: make(chan struct{}, 1),
+		closed:        closed,
+	}
+}
+
+// NextFrame returns a writer to write the next frame.
+func (e *encoder) NextFrame(ft frame.Type) (io.WriteCloser, error) {
+	select {
+	case <-e.closed:
+		return nil, io.EOF
+	default:
+	}
+
+	if !atomic.CompareAndSwapInt32(&e.hasNonClosedFrame, 0, 1) {
+		return nil, ErrNonCloseFrame
+	}
+
+	ret := &frameWriter{
+		writer:            e.writer,
+		hasFramesChan:     e.hasFramesChan,
+		hasNonClosedFrame: &e.hasNonClosedFrame,
+	}
+
+	if ft == frame.Binary {
+		if err := ret.writer.WriteByte(binaryPrefix); err != nil {
+			return nil, err
+		}
+		ret.base64 = base64.NewEncoder(base64.StdEncoding, ret.writer)
+	}
+
+	return ret, nil
+}
+
+// WriteFramesTo writes finished frames to the writer w.
+func (e *encoder) WriteFramesTo(w io.Writer) error {
+	pingTimeout := e.pingTimeout - time.Since(e.lastPing)
+	select {
+	case <-e.hasFramesChan:
+	case <-time.After(pingTimeout):
+		e.lastPing = time.Now()
+		return ErrPingTimeout
+	case <-e.closed:
+		return io.EOF
+	}
+
+	if _, err := e.writer.WriteFinishedFrames(w); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WaitFrameClose waits for the current frame writer finishing.
+func (e *encoder) WaitFrameClose() {
+	if atomic.LoadInt32(&e.hasNonClosedFrame) == 0 {
+		return
+	}
+
+	<-e.hasFramesChan
+	select {
+	case e.hasFramesChan <- struct{}{}:
+	default: // if it already has frames, the chan is full and continues.
+	}
+}
+
+type frameWriter struct {
```

Comment:
we need separate frameWriter and encoder by two different file in one package

Reply:
See the comment of `decoder` and `frameReader`.